ci-automation/test.sh

```diff
@@ -0,0 +1,184 @@
+# 2 - Architecture to test.
+# 3 - version number to test.
+# 4 - output TAP file.
+# All following arguments specify test cases / test case patterns to run.
+
+set -euo pipefail
+
+# Download torcx package and manifest, add build cache URL to manifest
+#  so the docker.torcx-manifest-pkgs test can use it.
+function __prepare_torcx() {
+    local arch="$1"
+    local vernum="$2"
+    local workdir="$3"
+
+    copy_from_buildcache "images/${arch}/${vernum}/torcx/torcx_manifest.json" "${workdir}"
+
+    local docker_pkg
+    docker_pkg="$(basename \
+                        "$(jq -r ".value.packages[0].versions[0].locations[0].path" \
+                        ${workdir}/torcx_manifest.json)")"
+
+    # Add docker package URL on build cache to manifest
+    local docker_url="http://${BUILDCACHE_SERVER}/images/${arch}/${vernum}/torcx/${docker_pkg}"
+    jq ".value.packages[0].versions[0].locations += [{\"url\" : \"${docker_url}\"}]" \
+        "${workdir}/torcx_manifest.json" \
+        > "${workdir}/torcx_manifest_new.json"
+
+    mv "${workdir}/torcx_manifest.json" "${workdir}/torcx_manifest.json.original"
+    mv "${workdir}/torcx_manifest_new.json" "${workdir}/torcx_manifest.json"
+}
+# --
+
+function test_run() {
+    local arch="$1" ; shift
+    local image="$1"; shift
+
+    # default to all tests
+    if [ $# -le 0 ] ; then
+        set -- '*'
+    fi
+
+    local retries="${MAX_RETRIES:-20}"
+
+    source ci-automation/tapfile_helper_lib.sh
+    source ci-automation/ci_automation_common.sh
+    init_submodules
+
+    source sdk_container/.repo/manifests/version.txt
+    local vernum="${FLATCAR_VERSION}"
+    local docker_vernum
+    docker_vernum="$(vernum_to_docker_image_version "${vernum}")"
+
+    # Get SDK from either the registry or import from build cache
+    local sdk_version="${FLATCAR_SDK_VERSION}"
+    local sdk_name="flatcar-sdk-${arch}"
+    local docker_sdk_vernum="$(vernum_to_docker_image_version "${sdk_version}")"
+
+    docker_image_from_registry_or_buildcache "${sdk_name}" "${docker_sdk_vernum}"
+    local sdk_image="$(docker_image_fullname "${sdk_name}" "${docker_sdk_vernum}")"
+    echo "docker image rm -f '${sdk_image}'" >> ./ci-cleanup.sh
+
+    local work_dir="__TESTS__"
+    local tests_dir="${work_dir}/${image}"
+    mkdir -p "${tests_dir}"
+    echo "sudo rm -rf '${tests_dir}'" >> ci-cleanup.sh
+
+    local container_name="flatcar-tests-${arch}-${docker_vernum}-${image}"
+
+    # Make the torcx artifacts available to test implementation
+    __prepare_torcx "${arch}" "${vernum}" "${work_dir}"
+
+    local retry=""
+    local success=false
+    for retry in $(seq "${retries}"); do
+        local tapfile="results-run-${retry}.tap"
+        local failfile="failed-run-${retry}.txt"
+
+        # Ignore retcode since tests are flaky. We'll re-run failed tests and
+        #  determine success based on test results (tapfile).
+        set +e -o noglob
+        ./run_sdk_container -x ./ci-cleanup.sh \
```

Comment:
It would be good to make Jenkins' `bin/` folder of the workspace available to use the provided kola binaries. The SDK kola binary is likely outdated and is hard to update for bug fixes. I think a well-known path for outside-provided kola binaries should always take precedence over the internal kola binary of the SDK.

Reply:
I'd argue we should rather update the SDK container more often then - we can make this straightforward and even automate the process for new kola releases. Motivation of using the test suite from the SDK container is that the test suite is versioned alongside each release version, and release tests are reproducible.

Reply:
Note that updating kola in the SDK container does not necessarily mandate a full SDK rebuild, as it would for the cork SDK - we could base the updated SDK on an existing SDK container, and supply a respective `Dockerfile` that only updates `kola`.

For example, consider SDK 3139.0.0 needs a `kola` update. We could provide a Dockerfile that builds a new SDK version 3139.1.0 based on the 3139.0.0 image, with an updated `kola` binary.

Reply:
With the current bind-mounts we could do something like `mv WORKSPACE/bin __build__/images/` to have the kola binaries available under `/mnt/host/source/src/build/bin` from inside the SDK

Reply:
> I'd argue we should rather update the SDK container more often then

The main problem is that we don't want to touch released SDKs which we would have to do in the two cases: 1. during a release we finished building everything and just have to fix a test case and your proposed way requires us to update the SDK image manually to do so. 2. the test's URLs for e.g., the kubeadm test change and the test fail due to this and we have to update the released SDKs now for users to be able to run the test suite successfully

Reply:
Really like the idea of injecting newer kola versions but I'd like to make it visible for users. Currently we don't. We also don't test new kola releases with old OS images, so there's likely bit-rot in that direction, too.

I wouldn't touch the SDKs once released, but we should make it easy for ourselves to supply an updated SDK for new (minor version and/or patch level) releases. I don't think this requires manually updating the SDK. Consider this:

```dockerfile
FROM ghcr.io/flatcar-linux/flatcar-sdk-all:3139.0.0
RUN ./update_chroot
```

being run with `docker build -t ghcr.io/flatcar-linux/flatcar-sdk-all:3139.1.0 .` would provide a new SDK with an updated kola (assuming the kola ebuild's `CROS_WORKON_COMMIT` points to the new version).

Reply:
Yes, it can be that some tests don't work on year old releases but for these cases one can use an older kola release (e.g., the one from the SDK at that time). In general we have the version filters to ensure that kola works with multiple release at the same time.

I don't follow the update_chroot example because the assumption that the ebuild file gets updated only works when the tag is forced pushed, or?

Reply:
> I don't follow the update_chroot example because the assumption that the ebuild file gets updated only works when the tag is forced pushed, or?

Force-push would not be necessary since we'd be releasing a new SDK (with a minor version / patch level version bump).

Trying to address the 2 cases you've stated:
> 1. during a release we finished building everything and just have to fix a test case and your proposed way requires us to update the SDK image manually to do so.
> 2. the test's URLs for e.g., the kubeadm test, change (or we have other fixes for the test because it is flaky etc) and the test fail due to this and we have to update the released SDKs now for users to be able to run the test suite successfully

In both cases I would release a new SDK based on an existing container, with minor / patch version bump. My argument is that this could be done automatically, and it should be done to keep the release version in sync with the environment used to build and to test it.

I've worked on a script which can be used to automate this: https://github.com/flatcar-linux/scripts/pull/230

Reply:
I've elaborated on the scenarios here: https://github.com/flatcar-linux/scripts/pull/230#issuecomment-1047612468

Reply:
> We also don't test new kola releases with old OS images, so there's likely bit-rot in that direction, too.

Actually, this is exactly what we do. All our release channels (including lts) use the current kola build.

Reply:
The response above was to an argument where users could reproduce release tests of old, released, OS images with new, updated kola versions. This is not guaranteed to work since we do not test this.